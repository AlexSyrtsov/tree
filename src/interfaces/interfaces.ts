export interface ICategory {
  category_name: string;
  category_id: string;
}

export interface IBrand {
  brand_name: string;
  brand_id: string;
}

export interface IModel {
  model_name: string;
  _id: string;
  varients: Array<IVariant>;
}

export interface IVariant {
  varient: string;
  current_price: number;
}

export interface ITreeItemData {
  title: string;
  id: string;
  parentId?: string;
  childrenIds?: Array<string>;
  subtitle?: string;
  disabled?: boolean;
  checked?: boolean;
  tagTitle?: string;
}
