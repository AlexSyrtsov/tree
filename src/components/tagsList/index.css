.tree-tags {
  display: flex;
  flex-direction: column;
  max-height: 50vh;
}

.tree-tags__wrapper {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 10px;
  overflow: scroll;
}
