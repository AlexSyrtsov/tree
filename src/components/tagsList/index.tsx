import { memo } from "react";

import { ITreeItemData } from "../../interfaces/interfaces";
import { Tag } from "../tag";

import "./index.css";

interface ITagsListProps {
  tags: Array<ITreeItemData>;
}

export const TagsList = memo(({ tags }: ITagsListProps) => {
  return (
    <div className="tree-tags">
      <h4 className="tree__title">Selected variants</h4>
      <div className="tree-tags__wrapper">
        {tags.map((tag) => (
          <Tag key={tag.id} tag={tag} />
        ))}
      </div>
    </div>
  );
});
