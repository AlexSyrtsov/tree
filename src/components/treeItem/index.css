.tree-item__wrapper {
  position: relative;
}

.tree-item__name-wrapper {
  display: flex;
  align-items: center;
  margin: 5px 0;
}

.tree-item__name-wrapper_without-icon {
  margin-left: 15px;
}

.tree-item__collapse-icon {
  width: 15px;
  cursor: pointer;

  -webkit-transition: 0.3s ease-in-out;
  -moz-transition: 0.3s ease-in-out;
  -o-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
}

.tree-item__collapse-icon_up {
  -webkit-transform: rotate(180deg);
  -moz-transform: rotate(180deg);
  -o-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.tree-item__checkbox-wrapper {
  display: flex;
  align-items: center;
}

.tree-item__checkbox {
  margin: 3px;
}

.tree-item__label-wrapper {
  display: flex;
  flex-direction: column;
}

.tree-item__label-subtitle {
  font-size: 12px;
}

.tree-item__children-wrapper {
  margin-left: 17px;

  -webkit-transition: 0.3s ease-in-out;
  -moz-transition: 0.3s ease-in-out;
  -o-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
}

.tree-item__children-enter {
  opacity: 0;
}
.tree-item__children-enter-active {
  opacity: 1;
  transition: opacity 300ms ease-in-out 200ms;
}
.tree-item__children-exit {
  opacity: 1;
}
.tree-item__children-exit-active {
  opacity: 0;
  transition: opacity 300ms ease-in-out 200ms;
}
