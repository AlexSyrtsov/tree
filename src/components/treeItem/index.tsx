import { memo, useCallback, useState } from "react";
import { TransitionGroup, CSSTransition } from "react-transition-group";

import DownIcon from "../../assets/down.svg";
import { ITreeItemData } from "../../interfaces/interfaces";

import "./index.css";

interface ITreeItemProps {
  data: ITreeItemData;
  getValueFromData: (id: string) => ITreeItemData;
}

export const TreeItem = memo(({ data, getValueFromData }: ITreeItemProps) => {
  const [isCollapsed, setIsCollapsed] = useState(true);

  const hasChildren = data?.childrenIds?.length;

  const toggle = useCallback(() => setIsCollapsed(!isCollapsed), [isCollapsed]);

  return (
    <div className="tree-item__wrapper">
      <div
        className={`tree-item__name-wrapper tree-item__name-wrapper_${
          hasChildren ? "with-icon" : "without-icon"
        }`}
      >
        {hasChildren && (
          <img
            className={`tree-item__collapse-icon tree-item__collapse-icon_${
              isCollapsed ? "down" : "up"
            }`}
            src={DownIcon}
            onClick={toggle}
          />
        )}
        <label className="tree-item__checkbox-wrapper" htmlFor={data.id}>
          <input
            className="tree-item__checkbox"
            type="checkbox"
            id={data.id}
            name={data.id}
            disabled={data.disabled}
            checked={data.checked}
          />
          <div className="tree-item__label-wrapper">
            <span className="tree-item__label-title">{data.title}</span>
            {isCollapsed && (
              <span className="tree-item__label-subtitle">{data.subtitle}</span>
            )}
          </div>
        </label>
      </div>
      <TransitionGroup className="tree-item__children-wrapper">
        {!isCollapsed &&
          data.childrenIds?.map((item) => (
            <CSSTransition
              key={item}
              timeout={500}
              classNames="tree-item__children"
            >
              <TreeItem
                data={getValueFromData(item)}
                getValueFromData={getValueFromData}
              />
            </CSSTransition>
          ))}
      </TransitionGroup>
    </div>
  );
});
