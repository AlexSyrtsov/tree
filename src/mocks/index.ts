import { ITreeItemData } from "../interfaces/interfaces";

export const MOCK_TREE_ITEM_FIRST: ITreeItemData = {
  id: "0030d9c5-9f1b-431a-9080-03cc48dc117f",
  title: "Phones",
};

export const MOCK_TREE_ITEM_SECOND: ITreeItemData = {
  id: "0030d9c5-9f1b-431a-9080-03cc48dc117d",
  title: "Laptops",
};

export const MOCK_TREE_ITEM_THIRD: ITreeItemData = {
  id: "0030d9c5-9f1b-431a-9080-03cc48dc117z",
  title: "Desktop",
  parentId: "0030d9c5-9f1b-431a-9080-03cc48dc117f",
};

export const MOCK_TREE_ITEM_FOURTH: ITreeItemData = {
  id: "0030d9c5-9f1b-431a-9080-03cc48dc117a",
  title: "Watches",
  parentId: "0030d9c5-9f1b-431a-9080-03cc48dc117f",
};

export const MOCK_TREE_ITEM_DISABLED: ITreeItemData = {
  ...MOCK_TREE_ITEM_FIRST,
  disabled: true,
};

export const MOCK_TREE_ITEM_WITH_CHILDREN: ITreeItemData = {
  ...MOCK_TREE_ITEM_FIRST,
  childrenIds: [
    "0030d9c5-9f1b-431a-9080-03cc48dc117a",
    "0030d9c5-9f1b-431a-9080-03cc48dc117z",
  ],
};

export const MOCK_ITEMS_ARRAY = [
  MOCK_TREE_ITEM_WITH_CHILDREN,
  MOCK_TREE_ITEM_SECOND,
  MOCK_TREE_ITEM_THIRD,
  MOCK_TREE_ITEM_FOURTH,
];

export const MOCK_ITEMS_MAP = {
  "0030d9c5-9f1b-431a-9080-03cc48dc117f": MOCK_TREE_ITEM_WITH_CHILDREN,
  "0030d9c5-9f1b-431a-9080-03cc48dc117d": MOCK_TREE_ITEM_SECOND,
};
