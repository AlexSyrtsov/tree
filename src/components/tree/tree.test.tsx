import * as enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";

import { Tree } from ".";
import { ITreeItemData } from "../../interfaces/interfaces";

import { MOCK_ITEMS_ARRAY } from "../../mocks";
import { TagsList } from "../tagsList";
import { TreeItem } from "../treeItem";

enzyme.configure({ adapter: new Adapter() });

const callback = (selectedElements: Array<ITreeItemData>) =>
  console.log(selectedElements);

describe("Tree component render correctly", () => {
  it("Match snapshot", () => {
    expect(
      <Tree data={MOCK_ITEMS_ARRAY} onChange={callback} />
    ).toMatchSnapshot();
  });

  it("Render correct title", () => {
    const result = enzyme
      .shallow(<Tree data={MOCK_ITEMS_ARRAY} onChange={callback} />)
      .contains(<h4 className="tree__title">Browse products</h4>);
    expect(result).toBeTruthy();
  });

  it("Render correct children", () => {
    const result = enzyme.shallow(
      <Tree data={MOCK_ITEMS_ARRAY} onChange={callback} />
    );

    expect(result.find(TreeItem)).toHaveLength(2);
  });
});
