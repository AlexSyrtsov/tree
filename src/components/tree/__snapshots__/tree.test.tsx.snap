// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree component render correctly Match snapshot 1`] = `
<Memo
  data={
    Array [
      Object {
        "childrenIds": Array [
          "0030d9c5-9f1b-431a-9080-03cc48dc117a",
          "0030d9c5-9f1b-431a-9080-03cc48dc117z",
        ],
        "id": "0030d9c5-9f1b-431a-9080-03cc48dc117f",
        "title": "Phones",
      },
      Object {
        "id": "0030d9c5-9f1b-431a-9080-03cc48dc117d",
        "title": "Laptops",
      },
      Object {
        "id": "0030d9c5-9f1b-431a-9080-03cc48dc117z",
        "parentId": "0030d9c5-9f1b-431a-9080-03cc48dc117f",
        "title": "Desktop",
      },
      Object {
        "id": "0030d9c5-9f1b-431a-9080-03cc48dc117a",
        "parentId": "0030d9c5-9f1b-431a-9080-03cc48dc117f",
        "title": "Watches",
      },
    ]
  }
  onChange={[Function]}
/>
`;
